src/mono/wasm/runtime/assets.ts

```diff
@@ -181,52 +225,87 @@ async function start_asset_download_sources(asset: AssetEntry): Promise<AssetEnt
-    while (throttling) {
-        await throttling.promise;
-    }
-    try {
-        ++parallel_count;
-        if (parallel_count == runtimeHelpers.maxParallelDownloads) {
-            if (runtimeHelpers.diagnosticTracing)
-                console.debug("MONO_WASM: Throttling further parallel downloads");
-            throttling = createPromiseController<void>();
+function resolve_path(asset: AssetEntry, sourcePrefix: string): string {
+    let attemptUrl;
+    const assemblyRootFolder = runtimeHelpers.config.assemblyRootFolder;
+    if (!asset.resolvedUrl) {
+        if (sourcePrefix === "") {
+            if (asset.behavior === "assembly" || asset.behavior === "pdb")
+                attemptUrl = assemblyRootFolder + "/" + asset.name;
+            else if (asset.behavior === "resource") {
+                const path = asset.culture !== "" ? `${asset.culture}/${asset.name}` : asset.name;
+                attemptUrl = assemblyRootFolder + "/" + path;
+            }
+            else {
+                attemptUrl = asset.name;
+            }
+        } else {
+            attemptUrl = sourcePrefix + asset.name;
         }
-        return await start_asset_download_sources(asset);
+        attemptUrl = runtimeHelpers.locateFile(attemptUrl);
     }
-    catch (response: any) {
-        const isOkToFail = asset.isOptional || (asset.name.match(/\.pdb$/) && runtimeHelpers.config.ignorePdbLoadErrors);
-        if (!isOkToFail) {
-            const err: any = new Error(`MONO_WASM: download '${response.url}' for ${asset.name} failed ${response.status} ${response.statusText}`);
-            err.status = response.status;
-            throw err;
-        }
+    else {
+        attemptUrl = asset.resolvedUrl;
     }
-    finally {
-        --parallel_count;
-        if (throttling && parallel_count == runtimeHelpers.maxParallelDownloads - 1) {
-            if (runtimeHelpers.diagnosticTracing)
-                console.debug("MONO_WASM: Resuming more parallel downloads");
-            const old_throttling = throttling;
-            throttling = undefined;
-            old_throttling.promise_control.resolve();
+    return attemptUrl;
+}
+
+function download_resource(request: ResourceRequest): LoadingResource {
+    try {
+        if (typeof Module.downloadResource === "function") {
+            const loading = Module.downloadResource(request);
+            if (loading) return loading;
+        }
+        const options: any = {};
+        if (request.hash) {
+            options.integrity = request.hash;
         }
+        const response = runtimeHelpers.fetch_like(request.resolvedUrl!, options);
+        return {
+            name: request.name, url: request.resolvedUrl!, response
+        };
+    } catch (err) {
+        const response = <Response><any>{
+            ok: false,
+            url: request.resolvedUrl,
+            status: 500,
+            statusText: "ERR28: " + err,
```

Comment:
What is `ERR28`?

Reply:
We are numbering some of the errors so that they are easier to find, when they happen. Search `/ERR[0-9]+/` in the code
This one should have new number, tho.